File: Algorithms/002. Add Two Numbers/Solution.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
public class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        return add(l1,l2,true);
    }
    public ListNode add(ListNode l1, ListNode l2,boolean flag) {
        boolean a=l1==null;
        boolean b=l2==null;
        if (a&&b&&flag) {
            return null;
        }
        int val=(a?0:l1.val)+(b?0:l2.val)+(flag?0:1);
        int v=val%10;
        boolean f=(v==val);
        ListNode answer=new ListNode(v);
        answer.next=add(a?null:l1.next,b?null:l2.next,f);
        return answer;
    }
}

File: Algorithms/002. Add Two Numbers/Solution.rb
# Definition for singly-linked list.
# class ListNode
#     attr_accessor :val, :next
#     def initialize(val)
#         @val = val
#         @next = nil
#     end
# end

# @param {ListNode} l1
# @param {ListNode} l2
# @return {ListNode}
def add_two_numbers(l1, l2)
    return add(l1,l2,0)
end

def add(l1,l2,f)
    if not l1 and not l2 and f==0
        return nil
    end
    ans=f +(l1 ? l1.val : 0)+(l2 ? l2.val : 0)
    res=ListNode.new(ans%10)
    res.next=add(l1 ? l1.next : nil,l2 ? l2.next : nil,ans/10)
    return res
end